nomad/stream/event_publisher.go

```diff
@@ -0,0 +1,220 @@
+package stream
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/nomad/nomad/structs"
+
+	"github.com/hashicorp/go-hclog"
+)
+
+const (
+	DefaultTTL = 1 * time.Hour
+)
+
+type EventPublisherCfg struct {
+	EventBufferSize int64
+	EventBufferTTL  time.Duration
+	Logger          hclog.Logger
+	OnEvict         EvictCallbackFn
+}
+
+type EventPublisher struct {
+	// lock protects the eventbuffer
+	lock sync.Mutex
```

Comment:
This lock seems to be held when accessing subscribers as well in `Subscribe` function.  I assume that's needed?